### PR TITLE
modified ACR module to support create and delete for snapshot,pe, and volume.

### PR DESCRIPTION
**Pull request by ar-india**
- the ACR module code prior to this only allowed a user to modify acr for a given volume. but going by the rest api spec. acr can also be created/deleted for snapshot, pe.  Hence have revamped the acr logic to make it more generic and support creation/deletion for all the different type of objects.

for a given ansible task like below

    # if state is create, then create acl for given volume, fails if it exist or cannot create
    # if state is present, then create acl if not present, else success
    - name: Create ACR for volume
      hpe_nimble_access_control_record:
        hostname: "{{ hostname }}"
        username: "{{ username }}"
        password: "{{ password }}"
        volume: "{{ansible_default_ipv4['address']}}-{{ volume }}"
        #initiator_group: "{{ansible_default_ipv4['address']}}-{{ initiator_group }}"
        # chap_user: "{{ chap_user }}"
        state: "{{ state | default('present') }}"

Output will be :

[root@centos80x64tmp ansible]# ansible-playbook playbook/hpe_nimble_access_control_records_playbook.yml -i hosts --tags "create_acr"

PLAY [localhost] ************************************************************************

TASK [Create ACR for volume] ************************************************************
**`fatal: [localhost]: FAILED! => {"changed": false, "msg": "Access control record creation failed. Please provide one of them: initiator_group or snapshot or protocol endpoint.", "return_status": false}`**

PLAY RECAP ******************************************************************************
localhost                  : ok=2    changed=0    unreachable=0    failed=1    skipped=0    rescued=0    ignored=0
*****************************************************

So , for creating acr , initiator group is always a mandatory param and one among volume, snapshot, pe or chap user is also required.

ex:
- name: Create ACR for volume
      hpe_nimble_access_control_record:
        hostname: "{{ hostname }}"
        username: "{{ username }}"
        password: "{{ password }}"
        volume: "{{ansible_default_ipv4['address']}}-{{ volume }}"
        initiator_group: "{{ansible_default_ipv4['address']}}-{{ initiator_group }}"
        state: "{{ state | default('present') }}"

Output:

[root@centos80x64tmp ansible]# ansible-playbook playbook/hpe_nimble_access_control_records_playbook.yml -i hosts --tags "create_acr" -v
Using /etc/ansible/ansible.cfg as config file

PLAY [localhost] ************************************************************************

TASK [Create ACR for volume] ************************************************************
`changed: [localhost] => {"changed": true, "msg": "Successfully created access control record for initiator group '10.18.180.239-ig-ansiblecentos1' associated with volume '10.18.180.239-ansible-vol1'.", "return_status": true}`

PLAY RECAP ******************************************************************************
localhost                  : ok=3    changed=1    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0

[root@centos80x64tmp ansible]#

### CODE REVIEW

File: ansible_collection/hpe/nimble/plugins/modules/hpe_nimble_access_control_record.py [hunk 1]
```diff
@@ -72,7 +72,8 @@
     required: False
     type: str
     description:
-    - Name of the snapshot this access control record applies to.
+    - Name of the snapshot this access control record applies to. If this record applies to a VVol snapshot, this attribute is required.
+      Otherwise, this attribute is not meaningful and should not be specified.
```
Comment (rgcostea) [inline ansible_collection/hpe/nimble/plugins/modules/hpe_nimble_access_control_record.py]: vVols should only be managed by the VASA provider on array, never externally.  Let's drop the reference to protocol_endpoint above and the additional text you added here.

File: ansible_collection/hpe/nimble/plugins/modules/hpe_nimble_access_control_record.py [hunk 2]
```diff
@@ -127,64 +129,65 @@
...
-        else:
-            # if state is set to present, we pass
-            if state == "present":
-                return (True, False, f"Access control record for volume '{volume}' with initiator group '{initiator_group}' is already present.")
-        return (False, False, f"Access control record for volume '{volume}' with initiator group '{initiator_group}' cannot be created as it is already present.")
+        params = utils.remove_null_args(**kwargs)
+        if 'apply_to' in params:
+            if params['apply_to'] != 'pe' and 'vol_id' in params is False:
+                # volume should be provided
+                return (False, False, "Access control record creation failed. Please provide valid volume.")
```
Comment (rgcostea) [inline ansible_collection/hpe/nimble/plugins/modules/hpe_nimble_access_control_record.py]: Drop the pe check

File: ansible_collection/hpe/nimble/plugins/modules/hpe_nimble_access_control_record.py [hunk 3]
```diff
@@ -127,64 +129,65 @@
...
+        # check if params has atleast 2 params. initiator group is mandatory along with any of below.
+        if params.__len__() < 2:
+            return (False, False, "Access control record creation failed. Please provide atleast one of them: volume, snapshot, protocol endpoint or chap user.")
+
+        if 'snap_id' in params:
+            resp = client_obj.snapshots.get(vol_id=params['vol_id'], id=params['snap_id'])
+            obj_type = 'snapshot'
+        elif 'pe_id' in params:
+            resp = client_obj.protocol_endpoints.get(id=params['pe_id'])
+            obj_type = 'protocol endpoint'
```
Comment (rgcostea) [inline ansible_collection/hpe/nimble/plugins/modules/hpe_nimble_access_control_record.py]: And here

File: ansible_collection/hpe/nimble/plugins/modules/hpe_nimble_access_control_record.py [hunk 4]
```diff
@@ -127,64 +129,65 @@
...
+        **kwargs):
 
-    if utils.is_null_or_empty(volume):
-        return (False, False, "Access control record deletion failed. No volume name Provided.")
     if utils.is_null_or_empty(initiator_group):
         return (False, False, "Access control record deletion failed. No initiator group provided.")
+    params = utils.remove_null_args(**kwargs)
+
+    if params.__len__() != 1:
+        return (False, False, "Access control record deletion failed. Please provide one of them: volume, snapshot, protocol endpoint or chap user.")
```
Comment (rgcostea) [inline ansible_collection/hpe/nimble/plugins/modules/hpe_nimble_access_control_record.py]: Drop protocol endpoint

File: ansible_collection/hpe/nimble/plugins/modules/hpe_nimble_access_control_record.py [hunk 5]
```diff
@@ -278,18 +281,25 @@ def main():
...
+            initiator_group,
             apply_to=apply_to,
             chap_user_id=utils.get_chap_user_id(client_obj, chap_user),
             lun=lun,
             pe_id=utils.get_pe_id(client_obj, protocol_endpoint),
-            snap_id=utils.get_snapshot_id(client_obj, snapshot),
+            snap_id=utils.get_snapshot_id(client_obj, volume, snapshot),
+            initiator_group_id=utils.get_initiator_group_id(client_obj, initiator_group),
+            vol_id=utils.get_vol_id(client_obj, volume),
             pe_ids=pe_ids)
```
Comment (rgcostea) [inline ansible_collection/hpe/nimble/plugins/modules/hpe_nimble_access_control_record.py]: Drop pe_ids

File: ansible_collection/hpe/nimble/plugins/modules/hpe_nimble_access_control_record.py [hunk 6]
```diff
@@ -278,18 +281,25 @@ def main():
...
 
     elif state == "absent":
-        return_status, changed, msg = delete_acr(client_obj, initiator_group, volume)
+        return_status, changed, msg = delete_acr(
+            client_obj,
+            initiator_group,
+            vol_name=volume,
+            pe_name=protocol_endpoint,
```
Comment (rgcostea) [inline ansible_collection/hpe/nimble/plugins/modules/hpe_nimble_access_control_record.py]: Drop this.

**Review by ar-india — COMMENTED** [review]
please review

File: ansible_collection/hpe/nimble/plugins/module_utils/hpe_nimble.py [hunk 7]
```diff
@@ -236,13 +236,13 @@ def get_pe_id(client_obj, pe_name):
...
-    if is_null_or_empty(snap_name):
+def get_snapshot_id(client_obj, vol_name, snap_name):
+    if is_null_or_empty(vol_name) or is_null_or_empty(snap_name):
         return None
     else:
-        resp = client_obj.snapshots.get(name=snap_name)
+        resp = client_obj.snapshots.get(vol_name=vol_name, name=snap_name)
         if resp is None:
-            raise Exception(f"Invalid value for snapshot: {snap_name}")
+            raise Exception(f"No snapshot with name '{snap_name}' found for volume: {vol_name}.")
```
Comment (suneeth51) [inline ansible_collection/hpe/nimble/plugins/module_utils/hpe_nimble.py]: nit: `volume {vol_name}`   (Colon not required)

File: ansible_collection/hpe/nimble/plugins/module_utils/hpe_nimble.py [hunk 8]
```diff
@@ -281,3 +281,12 @@ def get_downstream_partner_id(client_obj, downstream_partner):
         if resp is None:
             raise Exception(f"Invalid value for downstream partner: {downstream_partner}")
```
Comment (suneeth51) [inline ansible_collection/hpe/nimble/plugins/module_utils/hpe_nimble.py]: nit: No colon required `downstream partner {downstream_partner}`

File: ansible_collection/hpe/nimble/plugins/module_utils/hpe_nimble.py [hunk 9]
```diff
@@ -281,3 +281,12 @@ def get_downstream_partner_id(client_obj, downstream_partner):
...
         return resp.attrs.get("id")
+
+def get_initiator_group_id(client_obj, ig_name):
+    if is_null_or_empty(ig_name):
+        return None
+    else:
+        resp = client_obj.initiator_groups.get(name=ig_name)
+        if resp is None:
+            raise Exception(f"Invalid value for initiator group: {ig_name}")
```
Comment (suneeth51) [inline ansible_collection/hpe/nimble/plugins/module_utils/hpe_nimble.py]: nit: same as previous comment - Colon not required

File: ansible_collection/hpe/nimble/plugins/modules/hpe_nimble_snapshot_collection.py [hunk 10]
```diff
@@ -144,7 +144,7 @@
     volcoll: "{{ volcoll | mandatory}}"
     description: "{{ description }}"
 
-- name: Delete Snapshot collection collection  (must be offline)
+- name: Delete snapshot collection collection  (must be offline)
```
Comment (suneeth51) [inline ansible_collection/hpe/nimble/plugins/modules/hpe_nimble_snapshot_collection.py]: nit: `collection` repeated twice

**Review by suneeth51 — APPROVED** [review]
lgtm..

File: ansible_collection/hpe/nimble/plugins/module_utils/hpe_nimble.py (hunk 7 above)
Comment (ar-india) [reply ansible_collection/hpe/nimble/plugins/module_utils/hpe_nimble.py]: k

File: ansible_collection/hpe/nimble/plugins/modules/hpe_nimble_access_control_record.py (hunk 3 above)
Comment (ar-india) [reply ansible_collection/hpe/nimble/plugins/modules/hpe_nimble_access_control_record.py]: ok

File: ansible_collection/hpe/nimble/plugins/modules/hpe_nimble_access_control_record.py (hunk 4 above)
Comment (ar-india) [reply ansible_collection/hpe/nimble/plugins/modules/hpe_nimble_access_control_record.py]: done

**Pull request merged**